apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt

```diff
@@ -313,15 +313,23 @@ private constructor(
      * Configure the [WebSocketNetworkTransport] to reconnect the websocket automatically when a network error
      * happens
      *
-     * @param webSocketReconnectWhen a function taking the error as a parameter and returning 'true' to reconnect
-     * automatically or 'false' to forward the error to all listening [Flow]
+     * @param webSocketReconnectWhen a function taking the error and attempt index (starting from zero) as parameters
+     * and returning 'true' to reconnect automatically or 'false' to forward the error to all listening [Flow].
+     * It is a suspending function, so it can be used to introduce delay before retry (e.g. backoff strategy).
      *
      * See also [subscriptionNetworkTransport] for more customization
      */
-    fun webSocketReconnectWhen(webSocketReconnectWhen: ((Throwable) -> Boolean)) = apply {
+    fun webSocketReconnectWhen(webSocketReconnectWhen: (suspend (Throwable, attempt: Long) -> Boolean)) = apply {
       this.webSocketReconnectWhen = webSocketReconnectWhen
     }
 
+    fun webSocketReconnectWhen(reconnectWhen: (suspend (Throwable) -> Boolean)?) = apply {
```

Comment:
Thanks for keeping this one to avoid a breaking change 🙏!  Would it be possible to keep the lambda non suspending for this one? That would ensure binary compatibility as well.

Also I think we can mark it as `@Deprecated`.
```suggestion
    @Deprecated("Use webSocketReconnectWhen(webSocketReconnectWhen: (suspend (Throwable, attempt: Long) -> Boolean))")
    @ApolloDeprecatedSince(v3_0_1)
    fun webSocketReconnectWhen(reconnectWhen: ((Throwable) -> Boolean)?) = apply {
```

(Same goes for WebSocketNetworkTransport.kt below)